### PR TITLE
Main CSM Skeleton PR

### PR DESCRIPTION
**Pull request by ikekilinc**

### CODE REVIEW

File: csm/CameraMan.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+<<<<<<< HEAD
```
Comment (diegomtzg) [inline csm/CameraMan.py]: I don't know why these were marked as merge conflicts but can you get rid of all of these lines?

File: csm/ImageSources.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
-import cv2
+# import cv2
```
Comment (diegomtzg) [inline csm/ImageSources.py]: Let's try to keep the code in the repo meant to be run on the Jetson, can you uncomment these?

File: csm/MotorMan.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    def opticalFlowToMotorDeltas(self, optical_flow):
+        """
+        Converts optical flow data (vectors) into deltas motor must
+        turn to re-center camera view.
+        """
+        deltaX, deltaY = 0,0
+
+        pass
+        
+        return deltaX, deltaY
```
Comment (diegomtzg) [inline csm/MotorMan.py]: I don't think these would have to be returned. Rather, they'd have to become signals, but I guess this is fine until Thor fills it out

File: csm/PerceptionMan.py [hunk 4]
```diff
@@ -23,6 +23,7 @@ def __init__(self, network='ssd-mobilenet-v2', threshold=0.3):
         self.tracker = cv2.TrackerCSRT_create()
         # self.tracker = cv2.TrackerMOSSE_create()
         # self.tracker = cv2.TrackerMedianFlow_create()
+        pass
```
Comment (diegomtzg) [inline csm/PerceptionMan.py]: What's this pass for?

File: csm/StorageMan.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
+import queue
+# import cv2
```
Comment (diegomtzg) [inline csm/StorageMan.py]: No commented out code pls

File: csm/StorageMan.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+                frame, self.width, self.height = self.recvFrameQ.get()
+                self.frames.append(frame)
+                print("StorageMan   : Frame received and stored.")
+
+        opPath = self.compileInternal()
+        print("StorageMan   : Frames compiled and output saved at: %s" % opPath)
+
+    def compileInternal(self):
+        outputPath, fps = self.compileQ.get()
+        # output = cv2.VideoWriter(outputPath, cv2.VideoWriter_fourcc(*'DIVX'), fps, (width, height))
```
Comment (diegomtzg) [inline csm/StorageMan.py:39]: Can  you add a comment here about the current state of things (i.e. is it a WIP or like a TODO and then things you need to figure out)?

File: csm/SystemMan.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+import PerceptionMan as PerceptionMan
+
+from CommsMan import CommsMan
+from StorageMan import StorageMan
+from MotorMan import MotorMan
+from CameraMan import CameraMan
+from PerceptionMan import PerceptionMan
+
+# from ImageSources import ImageSource, LocalVideo, StillImage, Camera
+# from PerceptionUtils import BoundingBox
```
Comment (diegomtzg) [inline csm/SystemMan.py]: I don't know if you need these here, if you do, could you uncomment them? If you don't, could you get rid of them?

File: csm/SystemMan.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        self.per = PerceptionMan(threshold=0.5)
+        self.cam = CameraMan("../testData/testOutput.mp4")
+        self.mot = MotorMan()
+        
+
+        self.sto = StorageMan()
+        self.comQ, self.comT = queue.Queue(maxsize=1), queue.Queue(maxsize=1)
+        self.com = CommsMan(self.comQ, self.comT)
+
+    def sendMessageToRemote(self, message):
```
Comment (diegomtzg) [inline csm/SystemMan.py]: I don't really think we need these one line helper methods, but I'm open to hearing why you think we would

File: csm/SystemMan.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+                # Pull and process message from CommsMan
+                msg = self.comQ.get()
+
+                print("Main         : received message: %s" % msg)
+
+                # If msg commands new target selection
+                if (msg == "Terminate"):
+                    break
+                elif (msg != ''):
+                    # Restart PerceptionMan.
```
Comment (diegomtzg) [inline csm/SystemMan.py]: Could you add explicit TODOs here for what's missing?

File: csm/CommsMan.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+    def terminateCommsMan(self):
+        """
+        External function to terminate CommsMan thread operations.
+        """
+        try:
+            self.terminateComms.put(-1)
+            return 0
+        except:
+            return -1
```
Comment (diegomtzg) [inline csm/CommsMan.py:42]: When would this happen? Could you document it here with comments?

File: csm/MotorMan.py [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+    def adjustOrientation(self, optical_flow):
```
Comment (diegomtzg) [inline csm/MotorMan.py]: I'm a bit confused about the need for this function here. Can you not adjust the orientation with the function above?

**Review by diegomtzg — CHANGES_REQUESTED** [review]

File: csm/PerceptionMan.py (hunk 4 above)
Comment (ikekilinc) [reply csm/PerceptionMan.py]: A relic from when I had to comment out your code to make Demo1 compile

File: csm/MotorMan.py (hunk 11 above)
Comment (ikekilinc) [reply csm/MotorMan.py]: Agreed, can simply this module

File: csm/MotorMan.py (hunk 3 above)
Comment (ikekilinc) [reply csm/MotorMan.py]: Good point, I'm fine with sticking simple unless needed otherwise

File: csm/ImageSources.py (hunk 2 above)
Comment (ikekilinc) [reply csm/ImageSources.py]: Relic of past demo

File: csm/CommsMan.py (hunk 10 above)
Comment (ikekilinc) [reply csm/CommsMan.py:42]: Good note, important to comment in, reminded me of an addition needed as well

File: csm/CameraMan.py (hunk 1 above)
Comment (ikekilinc) [reply csm/CameraMan.py]: 👍

File: csm/SystemMan.py (hunk 8 above)
Comment (ikekilinc) [reply csm/SystemMan.py]: This is to provide an interface for us when we want to interact directly with SystemMan without removing the abstraction of specific data structures/objects used. Maybe more importantly, limiting any interaction with SystemMan to take place through these interfaces will help if SystemMan's implementation were to change.

File: csm/SystemMan.py (hunk 9 above)
Comment (ikekilinc) [reply csm/SystemMan.py]: Todos refer to portions of code you previous included in this version of PerceptionMan. How specific would you need the TODOs to be?

**Review by diegomtzg — CHANGES_REQUESTED** [review]
So I noticed that you moved the perception files to the csm directory, any particular reason why? Or was this on accident?

File: csm/CameraMan.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
+# import cv2
+
+class CameraMan():
+
+    def __init__(self, sourcePath):
+        self.updateVideoObject(sourcePath)
+
+    def updateVideoObject(self, sourcePath):
```
Comment (diegomtzg) [inline csm/CameraMan.py]: What does this function do?

Comment (diegomtzg) [reply csm/CameraMan.py]: I already implemented most of the camera work in ImageSources, would this call something from there? If so, could you get rid of capture here and leave a TODO for me to fill it in?

File: csm/CommsMan.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
+import threading
+import queue
+import time
+
+class CommsMan():
+
+    def __init__(self, commsQueue, termCommsQueue):
```
Comment (diegomtzg) [inline csm/CommsMan.py:7]: Let's talk later about the general purpose of some of these things

**Review by diegomtzg — CHANGES_REQUESTED** [review]
Looking good, great work here. In addition to moving the perception files back to the perception directory, and a few comments here and there, I'd just say to document some of the main functions that aren't entirely self-documenting (what it does, what arguments it takes, its outputs, etc) so that working on this code in the future is easier

File: csm/SystemMan.py (hunk 9 above)
Comment (diegomtzg) [reply csm/SystemMan.py]: Oh I just meant something like "# TODO: Call perception function that does this or whatever", I'll get rid of them once I do it so nothing fancy

File: csm/SystemMan.py [hunk 14]
```diff
@@ -0,0 +1,139 @@
+import CommsMan as CommsMan
+import StorageMan as StorageMan
+import MotorMan as MotorMan
+import CameraMan as CameraMan
+import PerceptionMan as PerceptionMan
+
+from CommsMan import CommsMan
```
Comment (diegomtzg) [inline csm/SystemMan.py:1]: These are duplicates, right? No need for the imports on the top

File: csm/SystemMan.py (hunk 8 above)
Comment (diegomtzg) [reply csm/SystemMan.py]: Ah I see. Abstraction... I like it.

File: csm/CommsMan.py (hunk 13 above)
Comment (diegomtzg) [reply csm/CommsMan.py:7]: Make sure to document what these data structures are for <3

**Review by diegomtzg — APPROVED** [review]

**Pull request merged**